core/Pos/Binary/Class/TH.hs

```diff
@@ -209,183 +183,80 @@ deriveSimpleBi headTy constrs = do
-    --    case x of
-    --        val@Bar{} -> getSize (f1 val) + getSize (f2 val)
-    --        val@Baz{} -> getSize (f1 val))
-    matchVarSize :: MatchQ
-    matchVarSize = do
-        x <- newName "x"
-        let branches = map matchVarCons filteredConstrs
-        match wildP
-              (normalB $
-                 appE [|Store.VarSize|] $
-                   lam1E (varP x)
-                     [| $mbTagSize + $(caseE (varE x) branches) |])
-              []
+    encodeTag :: Int -> Q Exp
+    encodeTag ix = [| Bi.encode (ix :: $tagType) |]
 
-    -- Generate the following code:
-    --     val@Bar{} -> getSize (field1 val) + getSize (field2 val)
-    matchVarCons :: Cons -> MatchQ
-    matchVarCons Cons{..} = do
-        -- we assume that the constructor is filtered and has only Field
-        fieldNames <- mapM (fmap fst . fieldToPair) cFields
-        val <- newName $ if null fieldNames then "_" else "val"
-        match (asP val (recP cName [])) (body (varE val) fieldNames) []
-      where
-        body val fieldNames = normalB $
-            sumE [[| Bi.getSize ($(varE f) $val) |] | f <- fieldNames]
+    encodeField :: ExpQ -> Field -> Q Exp
+    encodeField val Field{..} = do
+        (fName, _) <- expToNameAndType fFieldAndType
+        [| Bi.encode ($(varE fName) $val) |]
+    encodeField _  (Unused _) = fail "Something went wrong: encode Unused field"
 
-    -- Get definition --
-    biGetExpr :: Q Exp
-    biGetExpr =  appE [| Bi.label shortNameTy |] $ case constrs of
+    -- Decode definition --
+    biDecodeExpr :: Q Exp
+    biDecodeExpr = case constrs of
         []     ->
-            failText $ sformat ("Attempting to peek type without constructors "%shown) headTy
-        [cons] ->
-            (biGetConstr cons) -- There is one constructor
+            failText $ sformat ("Attempting to decode type without constructors "%shown) headTy
+        [cons] -> do
+          let unused = mkName "_"
+          doE [ bindS (varP unused)  [| Cbor.decodeListLen |]
+              , noBindS (biDecodeConstr cons) -- There is one constructor
+              ]
         _      -> do
             let tagName = mkName "tag"
+            let unused  = mkName "_"
             let getMatch ix con = match (litP (IntegerL (fromIntegral ix)))
-                                        (normalB (biGetConstr con)) []
+                                        (normalB (biDecodeConstr con)) []
             let mismatchConstr =
                     match wildP (normalB
-                        [| Store.peekException ("Found invalid tag while getting " <> shortNameTy) |]) []
+                        [| fail $ toString ("Found invalid tag while decoding " <> shortNameTy) |]) []
             doE
-                [ bindS (varP tagName) [| Bi.get |]
+                -- Unfortunately we cannot do any size-checking here as the incoming length won't
+                -- be known a-priori as this is machine-generated code.
```

Comment:
Good catch, this is a relic of the past we forgot to fix. We should be indeed able to do this as once we are in each case branch we should also know the number of args of each constructor. I'll see if I can fix this properly.